In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("C:\\Users\\zahee\\Desktop\\Data Science\\Project ExcelR\\telecommunications_churn.csv")
df

account_length  voice_mail_plan  voice_mail_messages  day_mins  \
0                128                1                   25     265.1   
1                107                1                   26     161.6   
2                137                0                    0     243.4   
3                 84                0                    0     299.4   
4                 75                0                    0     166.7   
...              ...              ...                  ...       ...   
3328             192                1                   36     156.2   
3329              68                0                    0     231.1   
3330              28                0                    0     180.8   
3331             184                0                    0     213.8   
3332              74                1                   25     234.4   

      evening_mins  night_mins  international_mins  customer_service_calls  \
0            197.4       244.7                10.0                       1   
1            195.5       254.4                13.7                       1   
2            121.2       162.6                12.2                       0   
3             61.9       196.9                 6.6                       2   
4            148.3       186.9                10.1                       3   
...            ...         ...                 ...                     ...   
3328         215.5       279.1                 9.9                       2   
3329         153.4       191.3                 9.6                       3   
3330         288.8       191.9                14.1                       2   
3331         159.6       139.2                 5.0                       2   
3332         265.9       241.4                13.7                       0   

      international_plan  day_calls  day_charge  evening_calls  \
0                      0        110       45.07             99   
1                      0        123       27.47            103   
2                      0        114       41.38            110   
3                      1         71       50.90             88   
4                      1        113       28.34            122   
...                  ...        ...         ...            ...   
3328                   0         77       26.55            126   
3329                   0         57       39.29             55   
3330                   0        109       30.74             58   
3331                   1        105       36.35             84   
3332                   0        113       39.85             82   

      evening_charge  night_calls  night_charge  international_calls  \
0              16.78           91         11.01                    3   
1              16.62          103         11.45                    3   
2              10.30          104          7.32                    5   
3               5.26           89          8.86                    7   
4              12.61          121          8.41                    3   
...              ...          ...           ...                  ...   
3328           18.32           83         12.56                    6   
3329           13.04          123          8.61                    4   
3330           24.55           91          8.64                    6   
3331           13.57          137          6.26                   10   
3332           22.60           77         10.86                    4   

      international_charge  total_charge  churn  
0                     2.70         75.56      0  
1                     3.70         59.24      0  
2                     3.29         62.29      0  
3                     1.78         66.80      0  
4                     2.73         52.09      0  
...                    ...           ...    ...  
3328                  2.67         60.10      0  
3329                  2.59         63.53      0  
3330                  3.81         67.74      0  
3331                  1.35         57.53      

In [3]:
df.drop(['voice_mail_plan','day_charge','evening_charge','night_charge', 'international_charge'],axis=1, inplace=True)
df

account_length  voice_mail_messages  day_mins  evening_mins  night_mins  \
0                128                   25     265.1         197.4       244.7   
1                107                   26     161.6         195.5       254.4   
2                137                    0     243.4         121.2       162.6   
3                 84                    0     299.4          61.9       196.9   
4                 75                    0     166.7         148.3       186.9   
...              ...                  ...       ...           ...         ...   
3328             192                   36     156.2         215.5       279.1   
3329              68                    0     231.1         153.4       191.3   
3330              28                    0     180.8         288.8       191.9   
3331             184                    0     213.8         159.6       139.2   
3332              74                   25     234.4         265.9       241.4   

      international_mins  customer_service_calls  international_plan  \
0                   10.0                       1                   0   
1                   13.7                       1                   0   
2                   12.2                       0                   0   
3                    6.6                       2                   1   
4                   10.1                       3                   1   
...                  ...                     ...                 ...   
3328                 9.9                       2                   0   
3329                 9.6                       3                   0   
3330                14.1                       2                   0   
3331                 5.0                       2                   1   
3332                13.7                       0                   0   

      day_calls  evening_calls  night_calls  international_calls  \
0           110             99           91                    3   
1           123            103          103                    3   
2           114            110          104                    5   
3            71             88           89                    7   
4           113            122          121                    3   
...         ...            ...          ...                  ...   
3328         77            126           83                    6   
3329         57             55          123                    4   
3330        109             58           91                    6   
3331        105             84          137                   10   
3332        113             82           77                    4   

      total_charge  churn  
0            75.56      0  
1            59.24      0  
2            62.29      0  
3            66.80      0  
4            52.09      0  
...            ...    ...  
3328         60.10      0  
3329         63.53      0  
3330         67.74      0  
3331         57.53      0  
3332         77.01      0  

[3333 rows x 14 columns]

In [5]:
# Splitting into X and Y
x= df.iloc[:,:-1]
Y= df.iloc[:,-1]

In [6]:
# Standardizing the input variables
from sklearn.preprocessing import StandardScaler
scale=StandardScaler()
X=scale.fit_transform(x)

In [10]:
# Splitting into test and train data
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(X,Y, test_size=0.2, random_state=7)

# Grid Search CV

In [11]:
from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.metrics import confusion_matrix, accuracy_score

In [12]:
clf=SVC()
param_grid=[{'kernel':['rbf'], 'gamma': [100,50,0.5,0.1,0.01,0.001,0.0001],
            'C':[15,14,13,12,11,10,0.01,0.001,0.0001]}]
gsv=GridSearchCV(clf,param_grid,cv=10)
gsv.fit(x_train,y_train)

GridSearchCV(cv=10, estimator=SVC(),
             param_grid=[{'C': [15, 14, 13, 12, 11, 10, 0.01, 0.001, 0.0001],
                          'gamma': [100, 50, 0.5, 0.1, 0.01, 0.001, 0.0001],
                          'kernel': ['rbf']}])

In [13]:
gsv.best_params_,gsv.best_score_

({'C': 10, 'gamma': 0.1, 'kernel': 'rbf'}, 0.9242220720340176)

In [14]:
clf=SVC(C=10, gamma=0.1)
clf.fit(x_train,y_train)
y_pred_train=clf.predict(x_train)
acc= accuracy_score(y_train,y_pred_train)*100
print("accuarcy=", acc)
print(confusion_matrix(y_train,y_pred_train))

accuarcy= 98.83720930232558
[[2284    0]
 [  31  351]]


In [15]:
y_pred_test=clf.predict(x_test)
acc=accuracy_score(y_test,y_pred_test)*100
print("accuarcy=", acc)
print(confusion_matrix(y_test,y_pred_test))

accuarcy= 93.10344827586206
[[548  18]
 [ 28  73]]


### optimizing the model for imbalanced dataset

In [16]:
weights={0:1,1:6}
clf=SVC(C=10, gamma=0.1, class_weight=weights)
clf.fit(x_train,y_train)
y_pred_train=clf.predict(x_train)
acc= accuracy_score(y_train,y_pred_train)*100
print("accuarcy=", acc)
print(confusion_matrix(y_train,y_pred_train))

accuarcy= 99.54988747186798
[[2273   11]
 [   1  381]]


In [18]:
y_pred_test=clf.predict(x_test)
acc=accuracy_score(y_test,y_pred_test)*100
print("accuarcy=", acc)
print(confusion_matrix(y_test,y_pred_test))

accuarcy= 91.1544227886057
[[537  29]
 [ 30  71]]


### Thus, we get an accuracy of 91.15% after dealing with class imbalance